gms/gossiper.cc

```diff
@@ -1375,22 +1375,29 @@ future<> gossiper::do_gossip_to_live_member(gossip_digest_syn message, gms::inet
 
 future<> gossiper::do_gossip_to_unreachable_member(gossip_digest_syn message) {
     double live_endpoint_count = _live_endpoints.size();
-    double unreachable_endpoint_count = _unreachable_endpoints.size();
-    if (unreachable_endpoint_count > 0) {
-        /* based on some probability */
-        double prob = unreachable_endpoint_count / (live_endpoint_count + 1);
-        std::uniform_real_distribution<double> dist(0, 1);
-        double rand_dbl = dist(_random_engine);
-        if (rand_dbl < prob) {
+    std::vector<inet_address> endpoints = _unreachable_endpoints.empty() && _live_endpoints.empty()
+        ? boost::copy_range<std::vector<inet_address>>(_endpoint_state_map | boost::adaptors::map_keys | boost::adaptors::filtered([me = get_broadcast_address()] (const inet_address& ep) {
```

Comment:
Again, when unreachable_endpoints.empty() && _live_endpoints.empty(), _endpoint_state_map is probably empty too on the new bootstrapping. So this is not going to talk to all nodes as wanted. We can test by adding logs to see if this code path is executed. 

And more importantly, we do want to talk to all _live_endpoints in the first few rounds to speed up propagation even if _liv_endpoints is not empty.

Reply:
@asias I don't understand how will that speed up propagation.
To start with, _live_endpoints will be empty initially, after `reset_endpoint_state_map`.
Then, once it is populated, we send gossip to a live endpoints (in groups of 10) right after `_endpoints_to_talk_with` empties, i.e. in the 2nd round.

Let's just close this PR, and if you see a viable way that I don't see, it would be simpler if you submit a PR with your proposal...

Reply:
Is there a related GH issue? I remember you mentioned a fail dtest.

The idea is sending to all live nodes in each round instead of sending 1/10 of live nodes in each round to speed up the propagation during the first few rounds.

Reply:
> Is there a related GH issue? I remember you mentioned a fail dtest.

There's https://github.com/scylladb/scylladb/issues/15268
and previously https://github.com/scylladb/scylladb/issues/15185.

> The idea is sending to all live nodes in each round instead of sending 1/10 of live nodes in each round to speed up the propagation during the first few rounds.

Ah I see, that's a fairly small change then.

Reply:
Regarding https://github.com/scylladb/scylladb/pull/15271#issuecomment-1705807262:
Using the live_endpoints as you proposed is a nice improvement overall, but it doesn't address the initial problem of discovering the live endpoints initially, after the shadow round.
Why do you think it's wrong to use a hint from the shadow round?

Reply:
OK. Btw, do you manage to reproduce 15268 locally?

Yup.   I mentioned the change here:

https://github.com/scylladb/scylladb/pull/15271#issuecomment-1705807262

Reply:
> OK. Btw, do you manage to reproduce 15268 locally?

No.  I think that https://github.com/scylladb/scylladb/issues/12282 which I'm working on could address this.
Since the point there is to have the host_id available along with the endpoint ip address over RPC.
The exception would be during upgrade where the other nodes run the old version and won't send their host_id when gossiping to a node restarting with the new version or bootstrapping running the new version.

Reply:
> Why do you think it's wrong to use a hint from the shadow round?

Because I want to remove the legacy shadow round eventually. Right now, we are already using a verb to get info needed for the shadow round purpose. There is no need to mess up the main gossip state if we are using the verb. It's better to avoid having new code to depend on the shadow round.

Reply:
Then, I will mark the optimization low priority. Since it does not solve the issue you mentioned.

Reply:
yeah, ok.  so let's close this PR then.